File: py/pe/pe14.py
#!/usr/bin/env python
"""
Which starting number produces the longest Collatz chain?
"""

_cache = {1:1}

def _chain(n):
    if not _cache.get(n, 0):
        if n & 1:
            _cache[n] = 1 + _chain(3 * n + 1)
        else:
            _cache[n] = 1 + _chain(n >> 1)
    return _cache[n]

def _chain_count(n):
    c = 1
    while n > 1:
        if n & 1:
            n = 3 * n + 1
        else:
            n >>= 1
        c += 1
    return c

def pe14(limit=1000000):
    """
    >>> pe14()
    (837799, 525)
    """
    m, mi = 0, 0
    for i in range(3, limit, 2):
        c = _chain(i)
        if m < c:
            m, mi = c, i
    return(mi, m)

if __name__ == "__main__":
    import doctest
    doctest.testmod()

    try:
        while True:
            s = input('> ')
            n = int(s)
            print(pe14(n))
    except (SyntaxError, EOFError, KeyboardInterrupt, NameError):
        pass

File: cpp/pe/src/pe14.cc
#include "../include/pe14.h"

namespace pe {
  uint32_t count(uint32_t n)
  {
    uint32_t c = 1;
    while (n > 1) {
      if (n & 1) {
        n = 3 * n + 1;
      } else {
        n >>= 1;
      }
      ++c;
    }

    return c;
  }

  void pe14(uint32_t& index, uint32_t& cnt, uint32_t begin, uint32_t limit)
  {
    uint32_t m = 0;
    uint32_t mi = 0;
    if ((begin & 1) == 0) {
      uint32_t c = count(begin);
      if (m < c) {
        m = c;
        mi = begin;
      }
      ++begin;
    }

    for (uint32_t i = begin; i < limit; i += 2) {
      uint32_t c = count(i);
      if (m < c) {
        m = c;
        mi = i;
      }
    }

    index = mi;
    cnt = m;
  }
}